client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala

```diff
@@ -0,0 +1,349 @@
+    }
+
+    if (!lifecycleManager.reserveSlotsWithRetry(
+        applicationId,
+        shuffleId,
+        new util.HashSet(candidates.toSet.asJava),
+        newlyAllocatedLocations)) {
+      logError(s"[Update partition] failed for $shuffleId.")
+      replyFailure(StatusCode.RESERVE_SLOTS_FAILED)
+      return
+    }
+
+    val newMasterLocations =
+      newlyAllocatedLocations.asScala.flatMap {
+        case (workInfo, (masterLocations, slaveLocations)) =>
+          // Add all re-allocated slots to worker snapshots.
+          lifecycleManager.workerSnapshots(shuffleId).asScala
+            .get(workInfo)
+            .foreach { partitionLocationInfo =>
+              partitionLocationInfo.addMasterPartitions(shuffleId.toString, masterLocations)
+              lifecycleManager.updateLatestPartitionLocations(shuffleId, masterLocations)
+              partitionLocationInfo.addSlavePartitions(shuffleId.toString, slaveLocations)
+            }
+          // partition location can be null when call reserveSlotsWithRetry().
+          val locations = (masterLocations.asScala ++ slaveLocations.asScala.map(_.getPeer))
+            .distinct.filter(_ != null)
+          if (locations.nonEmpty) {
+            val changes = locations.map { partition =>
+              s"(partition ${partition.getId} epoch from ${partition.getEpoch - 1} to ${partition.getEpoch})"
+            }.mkString("[", ", ", "]")
+            logDebug(s"[Update partition] success for " +
+              s"shuffle ${Utils.makeShuffleKey(applicationId, shuffleId)}, succeed partitions: " +
+              s"$changes.")
+          }
+          locations
+      }
+    replySuccess(newMasterLocations.toArray)
+  }
+
+  private def reallocateChangePartitionRequestSlotsFromCandidates(
+      changePartitionRequests: List[ChangePartitionRequest],
+      candidates: List[WorkerInfo]): WorkerResource = {
+    val slots = new WorkerResource()
+    changePartitionRequests.foreach { partition =>
+      lifecycleManager.allocateFromCandidates(
+        partition.partitionId,
+        partition.epoch,
+        candidates,
+        slots)
+    }
+    slots
+  }
+
+  private def blacklistPartition(
+      shuffleId: Int,
+      oldPartition: PartitionLocation,
+      cause: StatusCode): Unit = {
+    // only blacklist if cause is PushDataFailMain
+    val failedWorker = new ConcurrentHashMap[WorkerInfo, (StatusCode, Long)]()
+    if (cause == StatusCode.PUSH_DATA_FAIL_MASTER && oldPartition != null) {
+      val tmpWorker = oldPartition.getWorker
+      val worker = lifecycleManager.workerSnapshots(shuffleId).keySet().asScala
+        .find(_.equals(tmpWorker))
+      if (worker.isDefined) {
+        failedWorker.put(worker.get, (StatusCode.PUSH_DATA_FAIL_MASTER, System.currentTimeMillis()))
+      }
+    }
+    if (!failedWorker.isEmpty) {
+      lifecycleManager.recordWorkerFailure(failedWorker)
+    }
+  }
+
+  private def workersNotBlacklisted(shuffleId: Int): List[WorkerInfo] = {
```

Comment:
I think this method could be inline. Because it only used once and causing confusion. Why should change partition manager care about worker and blacklist?

Reply:
Only used here so I changed, but from the logic, should not change, you are right.